code/iaas/cluster-logic/src/main/java/io/cattle/iaas/cluster/service/impl/ClusterManagerImpl.java

```diff
@@ -235,33 +263,397 @@ public boolean evaluate(Instance obj) {
+        serverCertGen.setNotBefore(new Date());
+        serverCertGen.setNotAfter(new Date(System.currentTimeMillis() + CertificateConstants.EXPIRATION_MILLIS.get()));
+        serverCertGen.setSubjectDN(serverCSR.getCertificationRequestInfo().getSubject());
+        serverCertGen.setPublicKey(serverCSR.getPublicKey());
+        serverCertGen.setSignatureAlgorithm("SHA1WithRSAEncryption");
+
+        serverCertGen.addExtension(X509Extensions.ExtendedKeyUsage, false, new ExtendedKeyUsage(
+                KeyPurposeId.id_kp_serverAuth));
+        GeneralNames subjectAltName = new GeneralNames(new DERSequence(new GeneralName[] {
+                new GeneralName(GeneralName.iPAddress, ipAddress.getAddress()),
+                new GeneralName(GeneralName.iPAddress, "127.0.0.1")
+        }));
+        serverCertGen.addExtension(X509Extensions.SubjectAlternativeName, false, subjectAltName);
+
+        X509Certificate serverCrt = serverCertGen.generateX509Certificate(caKeyPair.getPrivate());
+        String serverCert = SshKeyGen.writePemObject(serverCrt);
+
+        // create cert record
+        Map<Object, Object> data = new HashMap<>();
+        data.put(CERTIFICATE.NAME, CertificateConstants.SERVER);
+        data.put(CERTIFICATE.KEY, serverKey);
+        data.put(CERTIFICATE.CERT, serverCert);
+        data.put(CERTIFICATE.CERT_CHAIN, caCertRecord.getCert());
+        data.put(CERTIFICATE.ACCOUNT_ID, host.getAccountId());
+        serverCertRecord = objectManager.create(Certificate.class, objectManager.convertToPropertiesFor(Certificate.class, data));
+
+        // update host with cert id
+        host.setCertificateId(serverCertRecord.getId());
+        objectManager.persist(host);
+    }
+
+    @SuppressWarnings("deprecation")
+    private void createClusterCertificates(Host cluster) throws Exception {
+        Certificate caCertRecord = getOrCreateCaCertificates(cluster.getAccountId());
+
+        Certificate clusterCertRecord = objectManager.findOne(Certificate.class,
+                CERTIFICATE.ACCOUNT_ID, cluster.getAccountId(),
+                CERTIFICATE.NAME, CertificateConstants.CLUSTER,
+                CERTIFICATE.CERT_CHAIN, caCertRecord.getCert(),
+                CERTIFICATE.REMOVED, null);
+
+        if (clusterCertRecord == null) {
+            KeyPair caKeyPair = SshKeyGen.readKeyPair(caCertRecord.getKey());
+
+            KeyPair swarmKeyPair = SshKeyGen.generateKeyPair();
+            String swarmKey = SshKeyGen.writePemObject(swarmKeyPair.getPrivate());
+
+            Host managingHost = getManagingHost(cluster);
+            IpAddress ipAddress = clusterHostMapDao.getIpAddressForHost(managingHost.getId());
+
+            PKCS10CertificationRequest swarmCSR = new PKCS10CertificationRequest(
+                    "SHA256withRSA",
+                    new X500Principal("CN="),
+                    swarmKeyPair.getPublic(),
+                    null,
+                    swarmKeyPair.getPrivate());
```

Comment:
As the above, we should probably wrap it in transaction. It would make sense to extract this method, place it in Dao, and enable the transaction on it. As it looks very similar to the one called from the previous method